File: engineering/src/stores/rlAgent.ts
/**
 * RL Agent Pinia Store（ADR-017 阶段 8 p8-6：决策推理 + 训练控制）
 *
 * 对接后端 `python/app/api/v1/rl_agent.py` 6 个端点：
 *   - GET    /api/v1/rl-agent/versions                 列出 RL 策略版本（分页 + algorithm/active 过滤）
 *   - GET    /api/v1/rl-agent/versions/{version}       查询策略版本详情
 *   - POST   /api/v1/rl-agent/act                      直接决策（不走工作流）
 *   - GET    /api/v1/rl-agent/training/status          查询训练状态
 *   - POST   /api/v1/rl-agent/training/start           启动训练 Workflow
 *   - POST   /api/v1/rl-agent/training/stop            停止训练
 *
 * 设计要点：
 *   1. **离线 RL 优先**：v1 仅支持基于历史数据 + 仿真环境的离线 RL
 *   2. **SafetyShield 硬约束**：强制过滤违反安全约束的动作，不可被 RL 策略覆盖
 *   3. **PPO 算法**：默认策略算法，clipped objective + GAE 优势估计
 *   4. **动作向量约定**：4 维 delta（主轴转速/进给/切深/切宽），取值 [-1, 1]
 *   5. POST /act 不持久化决策结果，前端如需保存可走工作流 rl_act 任务类型
 *   6. 训练控制端点为"指令式"：start 创建 RUNNING 记录，stop 置为 STOPPING，
 *      实际训练循环由后台 worker 异步执行
 *   7. recommended_action.reasoning 显式提示"本动作仅供 CAM 验证层参考，
 *      实际加工需经持证操作员审核"
 *   8. 错误统一通过 extractErrorMessage 提取，存入 error state
 */
import { defineStore } from 'pinia'
import { ref, computed } from 'vue'
import http from '@/utils/http'
import { extractErrorMessage } from '@/utils/error-handler'
import { unwrap } from '@/utils/response'
import { API_CONFIG, buildApiPath } from '@/config/api'
import {
  type PolicyVersion,
  type RLActRequest,
  type RLActResponse,
  type TrainingStatusInfo,
  type TrainingStartRequest,
  type ListPolicyVersionsParams,
  type ListPolicyVersionsResponse,
  type TrainingStatus,
} from '@/contracts/rl_agent'
import {
  TRAINING_STATUS,
  isTerminalTrainingStatus,
} from '@/contracts/rl_agent'

// ---------------------------------------------------------------------------
// Store 局部类型（派生 / UI 辅助）
// ---------------------------------------------------------------------------

/** 策略版本列表分页信息。 */
interface PolicyVersionPaginationState {
  total: number
  limit: number
  offset: number
}

/** fetchVersions 入参（与后端 Query 参数对齐）。 */
interface FetchPolicyVersionsParams extends ListPolicyVersionsParams {}

// ---------------------------------------------------------------------------
// Store 定义
// ---------------------------------------------------------------------------

export const useRlAgentStore = defineStore(
  'rlAgent',
  () => {
    // ===== State：策略版本列表 =====

    /** RL 策略版本列表（list 端点返回的 items）。 */
    const versions = ref<PolicyVersion[]>([])
    /** 策略版本列表分页信息。 */
    const versionPagination = ref<PolicyVersionPaginationState>({
      total: 0,
      limit: 50,
      offset: 0,
    })

    // ===== State：当前版本详情 =====

    /** 当前查看的 RL 策略版本详情。 */
    const currentVersion = ref<PolicyVersion | null>(null)

    // ===== State：最近一次决策结果 =====

    /** 最近一次 RL 决策响应（含推荐动作 + 候选评估 + 策略元信息）。 */
    const lastAction = ref<RLActResponse | null>(null)

    // ===== State：训练状态 =====

    /** 当前 RL 训练状态（status / 进度 / 指标 / 时间戳）。 */
    const trainingStatus = ref<TrainingStatusInfo | null>(null)

    // ===== Loading 标志 =====

    const versionsLoading = ref(false) // GET /versions
    const versionLoading = ref(false) // GET /versions/{version}
    const acting = ref(false) // POST /act
    const trainingStatusLoading = ref(false) // GET /training/status
    const startingTraining = ref(false) // POST /training/start
    const stoppingTraining = ref(false) // POST /training/stop
    const error = ref<string | null>(null)

    // ===== Computed =====

    /** 是否有 RL 策略版本。 */
    const hasVersions = computed<boolean>(() => versions.value.length > 0)

    /** 当前激活的策略版本（is_active=true）。 */
    const activeVersion = computed<PolicyVersion | null>(() =>
      versions.value.find((v) => v.is_active) ?? null,
    )

    /** 版本列表当前页码（1-based）。 */
    const currentPage = computed<number>(() =>
      versionPagination.value.limit > 0
        ? Math.floor(versionPagination.value.offset / versionPagination.value.limit) + 1
        : 1,
    )

    /** 版本列表总页数。 */
    const totalPages = computed<number>(() =>
      versionPagination.value.limit > 0
        ? Math.ceil(versionPagination.value.total / versionPagination.value.limit)
        : 1,
    )

    /** 是否有任何加载操作进行中。 */
    const anyLoading = computed<boolean>(
      () =>
        versionsLoading.value ||
        versionLoading.value ||
        acting.value ||
        trainingStatusLoading.value ||
        startingTraining.value ||
        stoppingTraining.value,
    )

    /** 当前训练状态字符串（无记录时为 idle）。 */
    const currentTrainingStatus = computed<TrainingStatus>(
      () => trainingStatus.value?.status ?? TRAINING_STATUS.IDLE,
    )

    /** 是否正在训练中（RUNNING / STOPPING 均视为占用训练槽位）。 */
    const isTraining = computed<boolean>(
      () =>
        currentTrainingStatus.value === TRAINING_STATUS.RUNNING ||
        currentTrainingStatus.value === TRAINING_STATUS.STOPPING,
    )

    /** 训练是否处于终态（COMPLETED / FAILED）。 */
    const isTrainingTerminal = computed<boolean>(() =>
      isTerminalTrainingStatus(currentTrainingStatus.value),
    )

    /** 训练进度百分比 [0, 100]。 */
    const trainingProgress = computed<number>(() => {
      const info = trainingStatus.value
      if (!info || info.max_steps <= 0) return 0
      return Math.min(100, Math.round((info.current_step / info.max_steps) * 100))
    })

    /** 最近一次决策推荐的动作是否安全通过（无安全违反）。 */
    const lastActionIsSafe = computed<boolean>(() => {
      const action = lastAction.value?.recommended_action?.reasoning ?? ''
      return !action.includes('违反') && !action.includes('已回退')
    })

    // unwrap() 已提取到 @/utils/response.ts（消除 6 个 Store 的重复定义）

    // ===== Actions：版本管理 =====

    /**
     * 列出 RL 策略版本（分页 + algorithm/active 过滤）.
     * 对应 GET /api/v1/rl-agent/versions
     *
     * @param params - 查询参数（active_only / algorithm / limit / offset）
     * @returns 策略版本列表响应，失败返回 null
     */
    async function fetchVersions(
      params: FetchPolicyVersionsParams = {},
    ): Promise<ListPolicyVersionsResponse | null> {
      versionsLoading.value = true
      error.value = null
      try {
        const response = await http.get(
          buildApiPath(API_CONFIG.RL_AGENT, '/versions'),
          {
            params: {
              active_only: params.active_only ?? false,
              algorithm: params.algorithm ?? undefined,
              limit: params.limit ?? 50,
              offset: params.offset ?? 0,
            },
          },
        )
        const payload = unwrap<ListPolicyVersionsResponse>(response)
        versions.value = payload.items ?? []
        versionPagination.value = {
          total: payload.total ?? 0,
          limit: payload.limit ?? 50,
          offset: payload.offset ?? 0,
        }
        return payload
      } catch (e) {
        error.value = extractErrorMessage(e, '获取 RL 策略版本列表失败')
        return null
      } finally {
        versionsLoading.value = false
      }
    }

    /**
     * 查询 RL 策略版本详情.
     * 对应 GET /api/v1/rl-agent/versions/{version}
     *
     * @param version - 版本号（semver）
     * @returns 策略版本详情，失败返回 null
     */
    async function fetchVersion(
      version: string,
    ): Promise<PolicyVersion | null> {
      versionLoading.value = true
      error.value = null
      try {
        const response = await http.get(
          buildApiPath(API_CONFIG.RL_AGENT, `/versions/${encodeURIComponent(version)}`),
        )
        const payload = unwrap<PolicyVersion>(response)
        currentVersion.value = payload
        return payload
      } catch (e) {
        error.value = extractErrorMessage(e, '获取 RL 策略版本详情失败')
        return null
      } finally {
        versionLoading.value = false
      }
    }

    // ===== Actions：决策推理 =====

    /**
     * 执行 RL 决策（不走工作流，直接调用服务层）.
     * 对应 POST /api/v1/rl-agent/act
     *
     * 流程：
     *   1. 提交当前状态 + 候选动作集 + 优化目标 + 安全约束
     *   2. 服务层执行策略前向 + SafetyShield 硬约束过滤 + 候选评估
     *   3. 返回推荐动作 + 候选评估列表 + 策略元信息
     *
     * 工程约束：
     *   - candidate_actions 至少 1 个，每个含 4 个 delta 字段
     *   - safety_constraints 为空时后端使用默认值
     *   - 推荐动作的 reasoning 显式提示"仅供 CAM 验证层参考，
     *     实际加工需经持证操作员审核"
     *   - 决策结果不持久化，如需保存走工作流 rl_act 任务类型
     *
     * @param request - 决策请求（含 current_state / candidate_actions / optimization_target / safety_constraints / model_uri）
     * @returns 决策响应（含推荐动作 + 候选评估 + 策略元信息），失败返回 null
     */
    async function act(
      request: RLActRequest,
    ): Promise<RLActResponse | null> {
      acting.value = true
      error.value = null
      try {
        const response = await http.post(
          buildApiPath(API_CONFIG.RL_AGENT, '/act'),
          request,
        )
        const payload = unwrap<RLActResponse>(response)
        lastAction.value = payload
        return payload
      } catch (e) {
        error.value = extractErrorMessage(e, 'RL 决策失败')
        return null
      } finally {
        acting.value = false
      }
    }

    // ===== Actions：训练控制 =====

    /**
     * 查询当前 RL 训练状态.
     * 对应 GET /api/v1/rl-agent/training/status
     *
     * 返回字段：
     *   - status: 训练状态（idle / running / paused / completed / failed / stopping）
     *   - current_step / max_steps / current_episode
     *   - metrics: 最新训练指标快照（仅 RUNNING 时返回）
     *   - started_at / finished_at / error_message
     *
     * 若无训练记录，返回 status=idle.
     *
     * @returns 训练状态信息，失败返回 null
     */
    async function fetchTrainingStatus(): Promise<TrainingStatusInfo | null> {
      trainingStatusLoading.value = true
      error.value = null
      try {
        const response = await http.get(
          buildApiPath(API_CONFIG.RL_AGENT, '/training/status'),
        )
        const payload = unwrap<TrainingStatusInfo>(response)
        trainingStatus.value = payload
        return payload
      } catch (e) {
        error.value = extractErrorMessage(e, '获取 RL 训练状态失败')
        return null
      } finally {
        trainingStatusLoading.value = false
      }
    }

    /**
     * 启动 RL 训练 Workflow.
     * 对应 POST /api/v1/rl-agent/training/start
     *
     * 流程：
     *   1. Pydantic 自动校验 max_steps / algorithm / optimization_target
     *   2. 服务层创建 RUNNING 记录
     *   3. 后台 worker 异步执行训练循环（v1 占位：实际训练循环由
     *      app.plugins.rl_agent.training.PPOTrainer 驱动）
     *
     * 工程约束：
     *   - 若已有 RUNNING 训练，后端返回 training_already_running 错误
     *   - v1 仅离线 RL：训练数据来自历史数据 + 仿真环境
     *   - 物理执行需"持证操作员 + 导师签字 + 保险"，本端点不涉及
     *
     * @param request - 启动训练请求（含 max_steps / seed / algorithm / optimization_target）
     * @returns 训练状态信息（RUNNING），失败返回 null
     */
    async function startTraining(
      request: TrainingStartRequest,
    ): Promise<TrainingStatusInfo | null> {
      startingTraining.value = true
      error.value = null
      try {
        const response = await http.post(
          buildApiPath(API_CONFIG.RL_AGENT, '/training/start'),
          request,
        )
        const payload = unwrap<TrainingStatusInfo>(response)
        trainingStatus.value = payload
        return payload
      } catch (e) {
        error.value = extractErrorMessage(e, '启动 RL 训练失败')
        return null
      } finally {
        startingTraining.value = false
      }
    }

    /**
     * 停止 RL 训练.
     * 对应 POST /api/v1/rl-agent/training/stop
     *
     * 将当前 RUNNING 训练置为 STOPPING，训练循环会在下一个 checkpoint
     * 边界退出并保存快照。
     *
     * @returns 训练状态信息（STOPPING），失败返回 null
     */
    async function stopTraining(): Promise<TrainingStatusInfo | null> {
      stoppingTraining.value = true
      error.value = null
      try {
        const response = await http.post(
          buildApiPath(API_CONFIG.RL_AGENT, '/training/stop'),
        )
        const payload = unwrap<TrainingStatusInfo>(response)
        trainingStatus.value = payload
        return payload
      } catch (e) {
        error.value = extractErrorMessage(e, '停止 RL 训练失败')
        return null
      } finally {
        stoppingTraining.value = false
      }
    }

    // ===== 清理方法 =====

    /** 清空当前版本详情。 */
    function clearCurrentVersion(): void {
      currentVersion.value = null
    }

    /** 清空最近一次决策结果。 */
    function clearLastAction(): void {
      lastAction.value = null
    }

    /** 清空训练状态。 */
    function clearTrainingStatus(): void {
      trainingStatus.value = null
    }

    /** 重置 Store 到初始状态。 */
    function $reset(): void {
      versions.value = []
      versionPagination.value = { total: 0, limit: 50, offset: 0 }
      currentVersion.value = null
      lastAction.value = null
      trainingStatus.value = null
      versionsLoading.value = false
      versionLoading.value = false
      acting.value = false
      trainingStatusLoading.value = false
      startingTraining.value = false
      stoppingTraining.value = false
      error.value = null
    }

    return {
      // State
      versions,
      versionPagination,
      currentVersion,
      lastAction,
      trainingStatus,
      versionsLoading,
      versionLoading,
      acting,
      trainingStatusLoading,
      startingTraining,
      stoppingTraining,
      error,
      // Computed
      hasVersions,
      activeVersion,
      currentPage,
      totalPages,
      anyLoading,
      currentTrainingStatus,
      isTraining,
      isTrainingTerminal,
      trainingProgress,
      lastActionIsSafe,
      // Actions
      fetchVersions,
      fetchVersion,
      act,
      fetchTrainingStatus,
      startTraining,
      stopTraining,
      // 清理方法
      clearCurrentVersion,
      clearLastAction,
      clearTrainingStatus,
      $reset,
    }
  },
)

File: engineering/python/app/api/v1/rl_agent.py
"""RL Agent API - 强化学习决策与训练控制 REST 接口.

对应 ADR-017（世界模型与 RL 模块）第 8 节：RL Agent REST API 端点。

端点总览（prefix: ``/api/v1/rl-agent``）：
    GET    /versions                 列出 RL 策略版本（分页 + algorithm/active 过滤）
    GET    /versions/{version}       查询策略版本详情
    POST   /act                      直接决策（不走工作流，调用 RLAgentService.act）
    GET    /training/status          查询训练状态
    POST   /training/start           启动训练 Workflow
    POST   /training/stop            停止训练

权限模型：
    rl_agent:read   —— 列出策略版本 / 查询版本详情 / 查询训练状态
    rl_agent:write  —— 直接决策 / 启动训练 / 停止训练

设计说明
--------
    - ``POST /act`` 不持久化决策结果，仅返回推荐动作 + 候选评估 + 策略元信息；
      前端如需持久化可走工作流 ``rl_act`` 任务类型
    - 训练控制端点为"指令式"：``start`` 创建 RUNNING 记录，``stop`` 将其置为
      STOPPING，实际训练循环由后台 worker 异步执行
    - 安全约束 ``SafetyConstraintsSpec`` 通过请求体传入，未传则使用默认值
    - 服务层异常通过 ``_handle_service_exception`` 统一映射为 API 错误响应
    - 所有决策响应包含 ``reasoning`` 字段（自然语言推荐理由），强调
      "本动作仅供 CAM 验证层参考，实际加工需经持证操作员审核"
"""
from __future__ import annotations

import logging
from typing import Any, Optional

from fastapi import APIRouter, Depends, Query
from pydantic import BaseModel, Field

from app.auth.permissions import require_permission
from app.core.response import ErrorCode, error, success
from app.dependencies import get_rl_agent_service
from app.contracts.rl_agent import (
    OptimizationTarget,
    PolicyAlgorithm,
    PolicyError,
    PolicyNotFoundError,
    RLActError,
    RLActRequest,
    SafetyConstraintsSpec,
    SafetyViolationError,
    TrainingAlreadyRunningError,
    TrainingError,
    TrainingStartRequest,
    TrainingStatus,
)

logger = logging.getLogger(__name__)

# 骨架修复（2026-08-03 任务B）：原文件缺失 router/logger/域符号导入。
# 补齐骨架但保持未接入（main/router_registry 未引用本文件）。
router = APIRouter(prefix="/api/v1/rl-agent", tags=["RL Agent"])




# ---------------------------------------------------------------------------
# Pydantic 请求模型
# ---------------------------------------------------------------------------


class SafetyConstraintsModel(BaseModel):
    """安全约束规格（与 ``SafetyConstraintsSpec`` 对齐）."""

    max_chatter_probability: float = Field(
        default=0.3,
        ge=0.0,
        le=1.0,
        description="最大允许颤振概率 [0, 1]，默认 0.3",
    )
    max_tool_wear_increment: float = Field(
        default=0.01,
        gt=0.0,
        description="最大允许刀具磨损增量 (mm/步)，默认 0.01",
    )
    min_surface_quality: float = Field(
        default=0.8,
        ge=0.0,
        le=1.0,
        description="最小表面质量 [0, 1]，默认 0.8",
    )

    def to_spec(self) -> SafetyConstraintsSpec:
        """转换为契约层 dataclass."""
        return SafetyConstraintsSpec(
            max_chatter_probability=self.max_chatter_probability,
            max_tool_wear_increment=self.max_tool_wear_increment,
            min_surface_quality=self.min_surface_quality,
        )


class RLActRequestModel(BaseModel):
    """RL 决策请求体.

    与 ``app.contracts.rl_agent.RLActRequest`` 对齐，但使用 Pydantic
    以获得自动校验和 OpenAPI 文档。
    """

    current_state: dict[str, float] = Field(
        ...,
        description="当前加工状态（字段名见 StateField，至少包含全部 8 个状态字段）",
    )
    candidate_actions: list[dict[str, float]] = Field(
        ...,
        min_length=1,
        description="候选动作集（至少 1 个，每个动作含 4 个 delta 字段）",
    )
    optimization_target: str = Field(
        default=OptimizationTarget.BALANCE,
        description=f"优化目标（{OptimizationTarget.all()}，默认 balance）",
    )
    safety_constraints: Optional[SafetyConstraintsModel] = Field(
        default=None,
        description="安全约束规格（为空则使用默认值）",
    )
    model_uri: str = Field(
        default="model://rl_agent/1.0.0",
        min_length=1,
        max_length=256,
        description="RL 策略模型 URI",
    )


class TrainingStartRequestModel(BaseModel):
    """启动训练请求体.

    与 ``app.contracts.rl_agent.TrainingStartRequest`` 对齐。
    """

    max_steps: int = Field(
        default=100000,
        ge=1000,
        le=1_000_000,
        description="最大训练步数（1000 ~ 1000000，默认 100000）",
    )
    seed: Optional[int] = Field(
        default=None,
        ge=0,
        le=2**31 - 1,
        description="随机种子（为空则使用训练器默认 42）",
    )
    algorithm: str = Field(
        default=PolicyAlgorithm.PPO,
        description=f"策略算法（{PolicyAlgorithm.all()}，默认 ppo）",
    )
    optimization_target: str = Field(
        default=OptimizationTarget.BALANCE,
        description=f"优化目标（{OptimizationTarget.all()}，默认 balance）",
    )


# ---------------------------------------------------------------------------
# 辅助函数
# ---------------------------------------------------------------------------


def _handle_service_exception(e: Exception, *, action: str):
    """统一处理服务层异常 → API 错误响应.

    风格与 explainability.py / world_model.py 对齐。

    Args:
        e: 服务层抛出的异常
        action: 当前操作描述（用于日志）

    Returns:
        error() 响应对象
    """
    if isinstance(e, PolicyNotFoundError):
        return error(
            code=ErrorCode.NOT_FOUND,
            message=str(e),
            suggestion="请确认 model_uri 正确，或通过 GET /versions 查看可用策略版本",
        )
    if isinstance(e, SafetyViolationError):
        return error(
            code=ErrorCode.INVALID_REQUEST,
            message=str(e),
            suggestion="所有候选动作均违反安全约束，请放宽 safety_constraints 或提供更多候选动作",
            recoverable=True,
        )
    if isinstance(e, TrainingAlreadyRunningError):
        return error(
            code=ErrorCode.INVALID_REQUEST,
            message=str(e),
            suggestion="已有训练任务运行中，请先调用 POST /training/stop 停止当前训练",
            recoverable=True,
        )
    if isinstance(e, PolicyError):
        return error(
            code=ErrorCode.INTERNAL_ERROR,
            message=str(e),
            suggestion="策略推理失败：模型权重可能未加载或维度不匹配，请检查 model_uri",
        )
    if isinstance(e, TrainingError):
        return error(
            code=ErrorCode.INTERNAL_ERROR,
            message=str(e),
            suggestion="训练控制失败：请检查训练配置或查看后端日志",
        )
    if isinstance(e, ValueError):
        return error(code=ErrorCode.INVALID_REQUEST, message=str(e))
    if isinstance(e, RLActError):
        logger.error("RLAgent error during %s: %s", action, e, exc_info=True)
        return error(code=ErrorCode.INTERNAL_ERROR, message=str(e))
    # 兜底：未识别的异常
    logger.error("Unexpected error during %s: %s", action, e, exc_info=True)
    return error(
        code=ErrorCode.INTERNAL_ERROR,
        message=f"{action} 失败",
        detail=str(e),
    )


# ---------------------------------------------------------------------------
# 端点 1: GET /versions —— 列出 RL 策略版本
# ---------------------------------------------------------------------------


@router.get("/versions")
async def list_versions(
    active_only: bool = Query(
        False, description="为 true 时仅返回当前激活版本"
    ),
    algorithm: Optional[str] = Query(
        None,
        description=f"按策略算法过滤（{PolicyAlgorithm.all()}）",
    ),
    limit: int = Query(
        50, ge=1, le=500, description="每页数量（1-500，默认 50）"
    ),
    offset: int = Query(0, ge=0, description="偏移量"),
):
    """分页列出 RL 策略版本.

    返回字段：
        - items: list[dict]（每个版本记录的 to_dict()）
        - total / limit / offset

    权限：``rl_agent:read``
    """
    # 前置校验：algorithm 合法性
    if algorithm is not None and not PolicyAlgorithm.is_valid(algorithm):
        return error(
            code=ErrorCode.INVALID_REQUEST,
            message=f"algorithm 不支持: {algorithm}"
            f"（支持: {PolicyAlgorithm.all()}）",
        )

    service = get_rl_agent_service()
    try:
        versions, total = await service.list_versions(
            active_only=active_only,
            algorithm=algorithm,
            limit=limit,
            offset=offset,
        )
    except Exception as e:
        return _handle_service_exception(e, action="列出 RL 策略版本")

    items = [v.to_dict() for v in versions]
    return success(
        data={
            "items": items,
            "total": total,
            "limit": limit,
            "offset": offset,
        },
        message=f"RL 策略版本列表已获取（{len(items)} 条）",
    )


# ---------------------------------------------------------------------------
# 端点 2: GET /versions/{version} —— 查询策略版本详情
# ---------------------------------------------------------------------------


@router.get("/versions/{version}")
async def get_version(version: str):
    """查询 RL 策略版本详情.

    权限：``rl_agent:read``
    """
    service = get_rl_agent_service()
    try:
        version_record = await service.get_version(version)
    except Exception as e:
        return _handle_service_exception(e, action="查询 RL 策略版本详情")

    return success(
        data=version_record.to_dict(),
        message="RL 策略版本详情已获取",
    )


# ---------------------------------------------------------------------------
# 端点 3: POST /act —— 直接决策（不走工作流）
# ---------------------------------------------------------------------------


@router.post(
    "/act",
    dependencies=[Depends(require_permission("rl_agent:write"))],
)
async def act(request: RLActRequestModel):
    """执行 RL 决策（不走工作流，直接调用服务层）.

    流程：
        1. Pydantic 自动校验 candidate_actions 非空 / optimization_target 合法
        2. 构造契约层 ``RLActRequest``（再次校验，与 Pydantic 互补）
        3. 调用 ``RLAgentService.act()`` 执行策略前向 + 安全过滤 + 候选评估
        4. 返回推荐动作 + 候选评估列表 + 策略元信息

    权限：``rl_agent:write``（触发模型推理，消耗资源）

    工程约束
    --------
        返回的 ``recommended_action.reasoning`` 字段会显式提示：
        "本动作仅供 CAM 验证层参考，实际加工需经持证操作员审核"
    """
    # 前置校验：optimization_target 合法性
    if not OptimizationTarget.is_valid(request.optimization_target):
        return error(
            code=ErrorCode.INVALID_REQUEST,
            message=f"optimization_target 不支持: {request.optimization_target}"
            f"（支持: {OptimizationTarget.all()}）",
        )

    # 解析安全约束（为空则使用默认值）
    if request.safety_constraints is not None:
        try:
            safety_spec = request.safety_constraints.to_spec()
        except ValueError as e:
            return error(
                code=ErrorCode.INVALID_REQUEST,
                message=f"safety_constraints 不合法: {e}",
            )
    else:
        safety_spec = SafetyConstraintsSpec()

    # 构造契约层 dataclass
    try:
        contract_req = RLActRequest(
            current_state=request.current_state,
            candidate_actions=request.candidate_actions,
            optimization_target=request.optimization_target,
            safety_constraints=safety_spec,
            model_uri=request.model_uri,
        )
    except ValueError as e:
        return error(
            code=ErrorCode.INVALID_REQUEST,
            message=str(e),
            suggestion="请检查 current_state / candidate_actions / optimization_target 字段",
        )

    service = get_rl_agent_service()
    try:
        response = await service.act(contract_req)
    except Exception as e:
        return _handle_service_exception(e, action="RL 决策")

    payload = response.to_dict()
    return success(
        data=payload,
        message=(
            f"RL 决策完成: 策略={response.policy_info.algorithm}，"
            f"候选数={len(response.action_evaluation)}，"
            f"推荐动作={'安全通过' if '违反' not in response.recommended_action.reasoning else '已回退'}"
        ),
    )


# ---------------------------------------------------------------------------
# 端点 4: GET /training/status —— 查询训练状态
# ---------------------------------------------------------------------------


@router.get("/training/status")
async def get_training_status():
    """查询当前 RL 训练状态.

    返回字段：
        - status: 训练状态（idle / running / paused / completed / failed / stopping）
        - current_step / max_steps / current_episode
        - metrics: 最新训练指标快照（仅 RUNNING 时返回）
        - started_at / finished_at / error_message

    若无训练记录，返回 status=idle.

    权限：``rl_agent:read``
    """
    service = get_rl_agent_service()
    try:
        status_info = await service.get_training_status()
    except Exception as e:
        return _handle_service_exception(e, action="查询 RL 训练状态")

    return success(
        data=status_info.to_dict(),
        message=f"训练状态: {status_info.status}",
    )


# ---------------------------------------------------------------------------
# 端点 5: POST /training/start —— 启动训练
# ---------------------------------------------------------------------------


@router.post(
    "/training/start",
    dependencies=[Depends(require_permission("rl_agent:write"))],
)
async def start_training(request: TrainingStartRequestModel):
    """启动 RL 训练 Workflow.

    流程：
        1. Pydantic 自动校验 max_steps / algorithm / optimization_target
        2. 构造契约层 ``TrainingStartRequest``
        3. 调用 ``RLAgentService.start_training()`` 创建 RUNNING 记录
        4. 后台 worker 异步执行训练循环（v1 占位：实际训练循环由
           ``app.plugins.rl_agent.training.PPOTrainer`` 驱动）

    权限：``rl_agent:write``

    工程约束
    --------
        - 若已有 RUNNING 训练，抛 ``TrainingAlreadyRunningError``
        - v1 仅离线 RL：训练数据来自历史数据 + 仿真环境
        - 物理执行需"持证操作员 + 导师签字 + 保险"，本端点不涉及
    """
    # 前置校验：algorithm 合法性
    if not PolicyAlgorithm.is_valid(request.algorithm):
        return error(
            code=ErrorCode.INVALID_REQUEST,
            message=f"algorithm 不支持: {request.algorithm}"
            f"（支持: {PolicyAlgorithm.all()}）",
        )
    # 前置校验：optimization_target 合法性
    if not OptimizationTarget.is_valid(request.optimization_target):
        return error(
            code=ErrorCode.INVALID_REQUEST,
            message=f"optimization_target 不支持: {request.optimization_target}"
            f"（支持: {OptimizationTarget.all()}）",
        )

    # 构造契约层 dataclass
    try:
        contract_req = TrainingStartRequest(
            max_steps=request.max_steps,
            seed=request.seed,
            algorithm=request.algorithm,
            optimization_target=request.optimization_target,
        )
    except ValueError as e:
        return error(
            code=ErrorCode.INVALID_REQUEST,
            message=str(e),
        )

    service = get_rl_agent_service()
    try:
        status_info = await service.start_training(contract_req)
    except Exception as e:
        return _handle_service_exception(e, action="启动 RL 训练")

    return success(
        data=status_info.to_dict(),
        message=(
            f"RL 训练已启动: algorithm={request.algorithm}，"
            f"max_steps={request.max_steps}"
        ),
    )


# ---------------------------------------------------------------------------
# 端点 6: POST /training/stop —— 停止训练
# ---------------------------------------------------------------------------


@router.post(
    "/training/stop",
    dependencies=[Depends(require_permission("rl_agent:write"))],
)
async def stop_training():
    """停止当前 RL 训练.

    流程：
        1. 查找 RUNNING 记录 → 置为 STOPPING
        2. 后台 worker 检测到 STOPPING 后保存 checkpoint 并退出
        3. 训练完成后状态变为 COMPLETED 或 FAILED

    若无 RUNNING 训练，返回当前状态（不报错）.

    权限：``rl_agent:write``
    """
    service = get_rl_agent_service()
    try:
        status_info = await service.stop_training()
    except Exception as e:
        return _handle_service_exception(e, action="停止 RL 训练")

    if status_info.status == TrainingStatus.IDLE:
        return success(
            data=status_info.to_dict(),
            message="当前无运行中的训练任务",
        )

    return success(
        data=status_info.to_dict(),
        message=f"已发送停止请求，训练状态: {status_info.status}",
    )


__all__ = ["router"]
